File: web/src/lib/engine/opciones.ts
/** Construcción de opciones inscribibles por curso — puerto de engine/opciones.py.
 *
 * 1. Filtra restringidas no habilitadas (con advertencia, nunca en silencio).
 * 2. Agrupa secciones equivalentes por (inicio, fin, días).
 * 3. Producto cartesiano clase × CADA componente práctico, descartando
 *    combinaciones internas con traslape (la trampa clase-choca-con-todo-lab).
 *
 * El ORDEN de las opciones resultantes es contrato: los tests de paridad y el
 * editor referencian opciones por índice.
 */
import {
  COMPONENTE_CLASE, minToHHMM, tieneHorario,
  type Curso, type Opcion, type Seccion, type Sesion,
} from "./models";
import { sesionesSeTraslapan } from "./overlap";

/** Agrupa secciones por (inicio, fin, dias) idénticos, preservando el orden
 * de aparición (igual que el dict de Python). */
export function agruparPorHorario(secciones: Seccion[]): Seccion[][] {
  const grupos = new Map<string, Seccion[]>();
  for (const sec of secciones) {
    const clave = `${sec.inicio_min}|${sec.fin_min}|${sec.dias.join(",")}`;
    const g = grupos.get(clave);
    if (g) g.push(sec);
    else grupos.set(clave, [sec]);
  }
  return [...grupos.values()];
}

function filtrar(
  secciones: Seccion[], incluirRestringidas: Set<string>,
  advertencias: string[], curso: Curso, nombreComp: string,
): Seccion[] {
  const conHorario = secciones.filter(tieneHorario);
  if (conHorario.length < secciones.length) {
    advertencias.push(
      `${curso.codigo}: ${secciones.length - conHorario.length} sección(es) de ` +
      `${nombreComp} sin horario definido en el catálogo; se ignoraron.`);
  }
  const elegibles = conHorario.filter(
    (s) => !s.restringida || incluirRestringidas.has(s.seccion));
  const fuera = conHorario
    .filter((s) => s.restringida && !incluirRestringidas.has(s.seccion))
    .map((s) => s.seccion);
  if (fuera.length && !elegibles.length) {
    advertencias.push(
      `${curso.codigo} (${nombreComp}): TODAS las secciones disponibles son ` +
      `restringidas (${fuera.join(", ")}). 'Ver Restricciones' no significa no ` +
      `disponible — si alguna aplica a tu carrera/pénsum, márcala como permitida.`);
  }
  return elegibles;
}

export function buildOpcionesCurso(
  curso: Curso, incluirRestringidas: Set<string> = new Set(),
): { opciones: Opcion[]; advertencias: string[] } {
  const advertencias: string[] = [];

  const componentes: Array<[string, Seccion[]]> = [];
  if (curso.secciones_clase.length) {
    componentes.push([COMPONENTE_CLASE, curso.secciones_clase]);
  }
  for (const [categoria, secs] of curso.componentes_practicos) {
    componentes.push([categoria, secs]);
  }

  const gruposPorComponente: Array<[string, Seccion[][]]> = [];
  for (const [nombreComp, secs] of componentes) {
    const elegibles = filtrar(secs, incluirRestringidas, advertencias, curso, nombreComp);
    const grupos = agruparPorHorario(elegibles);
    if (!grupos.length) return { opciones: [], advertencias };
    gruposPorComponente.push([nombreComp, grupos]);
  }

  const opciones: Opcion[] = [];
  const gruposClaseConSalida = new Set<Seccion[]>();

  // producto cartesiano en el MISMO orden que itertools.product
  const listas = gruposPorComponente.map(([, g]) => g);
  const indices = new Array(listas.length).fill(0);
  while (true) {
    const eleccion = listas.map((l, i) => l[indices[i]]);
    let valido = true;
    const comps = [];
    const sesiones: Sesion[] = [];
    for (let i = 0; i < eleccion.length; i++) {
      const grupo = eleccion[i];
      const rep = grupo[0];
      const sesion: Sesion = {
        inicio_min: rep.inicio_min!, fin_min: rep.fin_min!, dias: rep.dias,
      };
      if (sesionesSeTraslapan([sesion], sesiones)) { valido = false; break; }
      sesiones.push(sesion);
      comps.push({ categoria: gruposPorComponente[i][0], sesion, secciones: grupo });
    }
    if (valido) {
      opciones.push({ componentes: comps });
      gruposClaseConSalida.add(eleccion[0]);
    }
    // avanzar índices (último componente primero, como product)
    let k = listas.length - 1;
    while (k >= 0) {
      indices[k]++;
      if (indices[k] < listas[k].length) break;
      indices[k] = 0;
      k--;
    }
    if (k < 0) break;
  }

  // Advertencia explícita: clase incompatible con TODOS sus prácticos
  if (gruposPorComponente.length > 1) {
    for (const grupo of gruposPorComponente[0][1]) {
      if (!gruposClaseConSalida.has(grupo)) {
        const rep = grupo[0];
        const secs = grupo.map((s) => s.seccion).join("/");
        const practicos = gruposPorComponente.slice(1).map(([n]) => n).join(", ");
        advertencias.push(
          `${curso.codigo} sección ${secs} (${rep.dias.join(" ")} ` +
          `${minToHHMM(rep.inicio_min!)}–${minToHHMM(rep.fin_min!)}) se ` +
          `traslapa con TODAS las secciones de ${practicos}: es ` +
          `matemáticamente imposible de inscribir completa.`);
      }
    }
  }

  if (!opciones.length) {
    advertencias.push(
      `${curso.codigo}: ninguna combinación interna de clase + componentes ` +
      `prácticos es posible sin traslape.`);
  }
  return { opciones, advertencias };
}

File: engine/opciones.py
"""Construcción de opciones inscribibles por curso.

Pasos:
  1. Filtrar secciones restringidas que el usuario no marcó como aplicables
     (nunca se descartan en silencio: se genera una advertencia).
  2. Agrupar secciones equivalentes por (inicio, fin, días) — mismo horario,
     distinto catedrático.
  3. Producto cartesiano entre la clase y CADA componente práctico obligatorio
     (Laboratorio, Práctica, etc.), descartando combinaciones internas con
     traslape. Esto detecta trampas tipo "la clase choca con TODOS los labs".
"""
from __future__ import annotations

from itertools import product

from .models import COMPONENTE_CLASE, Componente, Curso, Opcion, Seccion, Sesion, min_to_hhmm, ordenar_dias
from .overlap import sesiones_se_traslapan


def agrupar_por_horario(secciones: list) -> list:
    """Agrupa secciones por (inicio, fin, dias) idénticos -> lista de grupos,
    cada grupo con sus secciones equivalentes (mismo horario, distinto profesor)."""
    grupos = {}
    for sec in secciones:
        key = (sec.inicio_min, sec.fin_min, sec.dias)
        grupos.setdefault(key, []).append(sec)
    return list(grupos.values())


def _filtrar(secciones: list, incluir_restringidas: set, advertencias: list,
             curso: Curso, nombre_comp: str) -> list:
    """Aplica los filtros de elegibilidad a un componente y explica qué quitó."""
    con_horario = [s for s in secciones if s.tiene_horario]
    if len(con_horario) < len(secciones):
        advertencias.append(
            f"{curso.codigo}: {len(secciones) - len(con_horario)} sección(es) de "
            f"{nombre_comp} sin horario definido en el catálogo; se ignoraron.")

    elegibles = [s for s in con_horario
                 if not s.restringida or s.seccion in incluir_restringidas]
    restringidas_fuera = [s.seccion for s in con_horario
                          if s.restringida and s.seccion not in incluir_restringidas]
    if restringidas_fuera and not elegibles:
        advertencias.append(
            f"{curso.codigo} ({nombre_comp}): TODAS las secciones disponibles son "
            f"restringidas ({', '.join(restringidas_fuera)}). 'Ver Restricciones' no "
            f"significa no disponible — si alguna aplica a tu carrera/pénsum, "
            f"márcala como permitida.")
    return elegibles


def build_opciones_curso(curso: Curso, incluir_restringidas: set = frozenset()):
    """Devuelve (opciones, advertencias) para un curso.

    Cada opción es un combo completo clase + un horario de cada componente
    práctico, ya validado internamente sin traslapes.
    """
    advertencias = []

    # Componentes en orden: clase primero (si existe), luego prácticos.
    # Un curso puede no tener clase magistral (ej. cursos que son solo lab).
    componentes = []
    if curso.secciones_clase:
        componentes.append((COMPONENTE_CLASE, curso.secciones_clase))
    for categoria, secs in curso.componentes_practicos.items():
        componentes.append((categoria, secs))

    grupos_por_componente = []
    for nombre_comp, secs in componentes:
        elegibles = _filtrar(secs, incluir_restringidas, advertencias, curso, nombre_comp)
        grupos = agrupar_por_horario(elegibles)
        if not grupos:
            # Sin este componente el curso no se puede inscribir completo.
            return [], advertencias
        grupos_por_componente.append((nombre_comp, grupos))

    opciones = []
    # Rastrea qué grupos de clase lograron al menos un combo válido, para
    # avisar explícitamente "esta sección choca con TODOS los laboratorios".
    grupos_clase_con_salida = set()

    for eleccion in product(*[grupos for _, grupos in grupos_por_componente]):
        comps, sesiones = [], []
        valido = True
        for (nombre_comp, _), grupo in zip(grupos_por_componente, eleccion):
            rep = grupo[0]
            sesion = Sesion(rep.inicio_min, rep.fin_min, rep.dias)
            if sesiones_se_traslapan([sesion], sesiones):
                valido = False
                break
            sesiones.append(sesion)
            comps.append(Componente(categoria=nombre_comp, sesion=sesion, secciones=grupo))
        if valido:
            opciones.append(Opcion(componentes=comps))
            grupos_clase_con_salida.add(id(eleccion[0]))

    # Advertencia explícita para clases incompatibles con todos sus prácticos
    # (la trampa que motivó este proyecto — ver SPEC.md sección 8).
    if len(grupos_por_componente) > 1:
        for grupo in grupos_por_componente[0][1]:
            if id(grupo) not in grupos_clase_con_salida:
                rep = grupo[0]
                secs = "/".join(s.seccion for s in grupo)
                dias = " ".join(ordenar_dias(rep.dias))
                practicos = ", ".join(n for n, _ in grupos_por_componente[1:])
                advertencias.append(
                    f"{curso.codigo} sección {secs} ({dias} "
                    f"{min_to_hhmm(rep.inicio_min)}–{min_to_hhmm(rep.fin_min)}) se "
                    f"traslapa con TODAS las secciones de {practicos}: es "
                    f"matemáticamente imposible de inscribir completa.")

    if not opciones:
        advertencias.append(
            f"{curso.codigo}: ninguna combinación interna de clase + componentes "
            f"prácticos es posible sin traslape.")
    return opciones, advertencias
